lib/python/rose/config_diff.py

```diff
@@ -0,0 +1,234 @@
+    for keylist in unset_keys:
+        config.unset(keylist)
+    return config
+
+
+def expand_shorthand_in_ignore_regexes(ignore_regexes_in):
+    """Expand any shorthand patterns in ignore_regexes."""
+    if ignore_regexes_in is None:
+        return None
+    ignore_regexes_in = list(ignore_regexes_in)
+    ignore_regexes = []
+    while ignore_regexes_in:
+        regex = ignore_regexes_in.pop(0)
+        is_shorthand = False
+        for shorthand, regexes in _DEFAULTS.SHORTHAND:
+            if shorthand == regex:
+                ignore_regexes_in.extend(regexes)
+                is_shorthand = True
+        if not is_shorthand:
+            ignore_regexes.append(regex)
+    return ignore_regexes
+
+
+def format_metadata_as_text(metadata, only_these_options=None):
+    """Convert a metadata dictionary to a block of text.
+
+    metadata -- a dictionary with metadata keys and values.
+    only_these_options -- (default None) if given, only output these
+    metadata keys. Otherwise, output all metadata keys.
+
+    """
+    id_node = rose.config.ConfigNode()
+    if only_these_options is None:
+        # Default to every option.
+        only_these_options = sorted(metadata.keys())
+    for property_ in only_these_options:
+        value = metadata.get(property_)
+        if value is None:
+            continue
+        id_node.set([property_], value=value)
+    string_file = StringIO.StringIO()
+    rose.config.dump(id_node, target=string_file)
+    string_file.seek(0)
+    return string_file.read()
+
+
+def main():
+    """Implement the "rose config-diff" command."""
+    opt_parser = rose.opt_parse.RoseOptionParser()
+    opt_parser.add_my_options("diff_tool", "graphical",
+                              "ignore", "meta_path", "properties")
+    my_sys_args = list(sys.argv)
+    if "--" in sys.argv:
+        my_sys_args = sys.argv[:sys.argv.index("--")]
+        diff_args = sys.argv[sys.argv.index("--") + 1:]
+    else:
+        diff_args = []
+
+    opts, args = opt_parser.parse_args(my_sys_args[1:])
+    rose.macro.add_site_meta_paths()
+    rose.macro.add_env_meta_paths()
+    rose.macro.add_opt_meta_paths(opts.meta_path)
+
+    if len(args) != 2:
+        sys.exit(opt_parser.get_usage())
+
+    config_loader = rose.config.ConfigLoader()
+
+    if opts.properties is None:
+        properties = _DEFAULTS.PROPERTIES.split(",")
+    else:
+        properties = opts.properties.split(",")
+
+    ignore_regexes = expand_shorthand_in_ignore_regexes(
+        opts.ignore_patterns)
+
+    output_filenames = []
+    config_type = rose.SUB_CONFIG_NAME
+    for path in args:
+        if path == "-":
+            continue
+        config_type = path.split(os.sep)[-1]
+    for path in args:
+        if path == "-":
+            path = sys.stdin
+            filename = config_type
+            directory = None
+        else:
+            filename = path.split(os.sep)[-1]
+            directory = path.rsplit(os.sep, 1)[0]
```

Comment:
Don't fancy using `os.path.basename` and `os.path.dirname`?

Alternatively, one can do:

```python
directory, filename = path.rsplit(os.sep, 1)
```